### PR TITLE
Improve client authentication and error handling

### PR DESCRIPTION
**Pull request by darinkrauss**
- Improve client authentication and error handling
- Use server session token provider rather than direct server session token with clients
- Separate platform-specific endpoint error handling from platform client
- Handle client response based upon client-specific error handler
- Update OAuth client to accept base client rather that constructing one
- Update OAuth refresh token error determination
- Fix stale server session tokens with long running tasks
- Update to latest
- Update tests

The current implementation of the various clients (based upon the root `client` package, and the `platform` client, in particular) had two significant issues:

1. Any client that relied on a server session token embedded in the context would sometimes fail if the server session token provided expired while the client was in use (failing with an "authentication token is invalid" error). This could happen occasionally with any client, but much more frequently happened with the Dexcom task clients (because doing a full fetch of a new Dexcom connection could take quite a while, giving a lot of time for the server session token to expire). The clients were all updated to use a server session token *provider* embedded in the context so *every* network request would use the *latest* server session token.

2. Whenever the `platform` client received a non-200 status code, it would *always* attempt to parse the response body for a `platform` `SerializedError` structure so that it could forward that error on. The `platform` client was used as the basis for other clients that didn't necessarily make network requests to `platform`-based services and the response body parsing would completely fail (either resulting in erroneous log messages) or, worse, if the response body included something parsable it generate a parsed error unrelated to the actual network error. The code was updated so the client can be configured to use an `ErrorParser` or not. For non-`platform`-based services there is a `NewLegacyClient` function to create a client that does not parse the response body. Furthermore, the handling of errors and capturing the response body was streamlined and fixed.

Also, some changes to the `oauth` client and the third-party OAuth library was updated.

NOTE: This is a part of the overall Dexcom work which will be collected into a final PR for final approval. However, since the work covers a number of different issues, I broke up the development into multiple smaller and more focused PRs. Once all of the smaller PRs are approved, I'll create a final overall PR for approval that will eventually be tested and deployed. (The smaller PRs will not be tested nor deployed.)

### CODE REVIEW

**Review by toddkazakov — APPROVED** [review]

File: auth/client/client_test.go [hunk 1]
```diff
@@ -32,14 +36,21 @@ var _ = Describe("Client", func() {
...
 	BeforeEach(func() {
+		serverSessionTokenProviderController = gomock.NewController(GinkgoT())
+		serverSessionTokenProvider = authTest.NewMockServerSessionTokenProvider(serverSessionTokenProviderController)
 		serverTokenSecret = authTest.NewServiceSecret()
 		serverTokenTimeout = testHttp.NewTimeout()
 		name = test.RandomStringFromRangeAndCharset(4, 16, test.CharsetAlphaNumeric)
 		logger = logTest.NewLogger()
 		Expect(logger).ToNot(BeNil())
 		serverToken = authTest.NewSessionToken()
+		serverSessionTokenProvider.EXPECT().ServerSessionToken().Return(serverToken, nil).AnyTimes()
```
Comment (jh-bate) [inline auth/client/client_test.go:47]: just a query `.EXPECT()` ?? I see its generated I just didn't expect (haha) that casing

File: ehr/reconcile/runner_test.go [hunk 2]
```diff
@@ -148,7 +148,6 @@ var _ = Describe("Runner", func() {
 					tasksList = append(tasksList, &t)
 				}
 
-				authClient.EXPECT().ServerSessionToken().Return("token", nil)
 				clinicsClient.EXPECT().ListEHREnabledClinics(gomock.Any()).Return(clinics, nil)
```
Comment (jh-bate) [inline ehr/reconcile/runner_test.go:151]: ahh ok - now I see it else where :)

**Review by jh-bate — APPROVED** [review]
LGTM and til about `EXPECT`

File: auth/client/client_test.go (hunk 1 above)
Comment (darinkrauss) [reply auth/client/client_test.go:47]: Yea, this is how https://github.com/golang/mock generates.

**Comment by darinkrauss** [comment]
Closing. Included in grouped PR https://github.com/tidepool-org/platform/pull/764.

**Pull request closed without merging**